### DISCUSSION TITLE
EMQX subscriptions API showing subscriptions of client not available

### DISCUSSION THREAD

**Discussion by dineshwar21** [emqx/emqx#14011] [Q&A]
I'm using the following API endpoint to retrieve all subscription topics:

`/api/v5/subscriptions`

However, I observe that many of the subscriptions listed are associated with clients whose connections have already been closed. Despite the disconnection, these subscriptions remain active and visible in the results. This behavior is unexpected, as I would expect subscriptions to be removed when the client disconnects.

![image](https://github.com/user-attachments/assets/06028901-5a49-4b7e-b9a0-bcbb32aad7ec)

I also checked if client exits as Persistent Session in client list, it seems no disconnected clients

Also reason for the client disconnect is 'ssl_closed'

**Comment by id** [comment]
@dineshwar21 please check if you have session expiry interval set by the clients, or on MQTT level in EMQX.

**Reply by dineshwar21** [reply]
![image](https://github.com/user-attachments/assets/1a46a462-ab16-4d9d-a23f-8e22a2ec42ef)

session expiry seems to be 0

**Reply by id** [reply]
And if you check parameters of one of the existing sessions in the dashboard?

**Reply by dineshwar21** [reply]
![image](https://github.com/user-attachments/assets/906a737f-3b2e-4d97-a489-1a4984e942a5)

Are you referring to this

**Reply by id** [reply]
Yes. Are you saying that this exact session is disconnected?

**Reply by dineshwar21** [reply]
I cannot find the client in API or in the dashboard, it seems clients as be disconnected and close. But subscription which it created still there. So subscriptions API response having orphan subscription which don't have a client.

**Reply by id** [reply]
Can you run this on EMQX node as `emqx` user?

```
emqx eval 'ets:lookup(emqx_channel, <<"mqttx_7d23c816">>).'
```

Instead of `mqttx_7d23c816` put real client id of one of your orphan subscriptions.

**Reply by dineshwar21** [reply]
@id 
Sorry do to heavy load, I have restarted EMQX service now orphan subscriptions removed. I'm not sure what cause orphan subscriptions.

**Reply by ieQu1** [reply]
Do you use core/replicant setup?

**Reply by dineshwar21** [reply]
@ieQu1 
If you are referring multiple cluster, we are currently using a single cluster only

**Reply by ieQu1** [reply]
I was referring to https://docs.emqx.com/en/emqx/latest/design/clustering.html#core-and-replicant-nodes

**Reply by dineshwar21** [reply]
No we have only one cluster @ieQu1

**Reply by ieQu1** [reply]
This feature works **within one cluster**, so maybe you mean "one node".

**Reply by dineshwar21** [reply]
Yes only one node.
![image](https://github.com/user-attachments/assets/3da1ee5e-e0af-4f9e-91fa-17d3df5569ee)

**Reply by dineshwar21** [reply]
@id 
The issue seems to be happening again.
```
$emqx eval 'ets:lookup(emqx_channel, <<"ti__8bSafgl6GYy12mLpjP2XI">>).'
[]
```
![image](https://github.com/user-attachments/assets/496a5181-520a-4aa7-bf52-d3296fbb17c8)
![image](https://github.com/user-attachments/assets/100b36d4-768a-4aa8-aad7-70bdd3af56b3)

**Comment by zmstone** [comment]
@dineshwar21  could you share the output of below commands:
- `emqx ctl conf show mqtt`
- `emqx ctl conf show durable_sessions`

**Reply by dineshwar21** [reply]
@zmstone 
$ emqx ctl conf show mqtt
mqtt {
  await_rel_timeout = "300s"
  client_attrs_init = []
  exclusive_subscription = false
  idle_timeout = "15s"
  ignore_loop_deliver = false
  keepalive_check_interval = "30s"
  keepalive_multiplier = 1.5
  max_awaiting_rel = 100
  max_clientid_len = 65535
  max_inflight = 32
  max_mqueue_len = 1000
  max_packet_size = "1MB"
  max_qos_allowed = 2
  max_subscriptions = infinity
  max_topic_alias = 65535
  max_topic_levels = 128
  message_expiry_interval = infinity
  mqueue_default_priority = lowest
  mqueue_priorities = disabled
  mqueue_store_qos0 = true
  peer_cert_as_clientid = disabled
  peer_cert_as_username = cn
  response_information = ""
  retain_available = true
  retry_interval = "30s"
  server_keepalive = disabled
  session_expiry_interval = "2h"
  shared_subscription = true
  shared_subscription_strategy = round_robin
  strict_mode = false
  upgrade_qos = false
  use_username_as_clientid = false
  wildcard_subscription = true
}

$ emqx ctl conf show durable_sessions
durable_sessions {
  batch_size = 100
  enable = false
  heartbeat_interval = "5000ms"
  idle_poll_interval = "100ms"
  message_retention_period = "1d"
  session_gc_batch_size = 100
  session_gc_interval = "10m"
}

**Comment by zmstone** [comment]
Hi @dineshwar21 Thank you for sharing the configs.
I cannot see any issue from the shared information so far.
And I cannot reproduce this issue.

Do you have a stable way to reproduce it? It would be great if you can share the steps with us.

**Reply by dineshwar21** [reply]
Hi @zmstone 
There is no stable way to reproduce it. It happening randomly.

**Comment by Khayhen** [comment]
I am facing a similar issue and manage to recreate it consistently. I have deployed a fresh EMQX instance with all default settings and has only set the authentication.
These are the metrics at start.
![image](https://github.com/user-attachments/assets/4a7f6c70-a312-4039-aa53-cb09a993d6a3)

I then deploy 6 clients that subscribe to 3446 topics which all of which are shared except 2. All the 6 clients are connected in a clean session mode. The metrics are as follow.
![image](https://github.com/user-attachments/assets/5e5ec243-ff8a-4d38-b75f-2f10f219b261)

Then I close the 6 clients. The metrics are as follow.
![image](https://github.com/user-attachments/assets/f8634e9c-edd5-4c5a-8c92-b3ba94a2cb90)

As you can see the subscriptions has not been removed but the topics has been removed. Hope you can help me solve this problem, thank you.

**Reply by keynslug** [reply]
Hi @dineshwar21, I can't quickly reproduce. Got a theory however, a couple of questions:
1. Do you have any errors / warning in the logs when this happens?
2. What hardware is it running on? Storage and IO capacity are of primary interest.
3. Just to be sure: are MQTT clients using _Session Expiry Interval_ equal to 0?
